src/app/Vitessce.js

```diff
@@ -0,0 +1,154 @@
+
+  // Emit the initialized view config if it is different than the
+  // prop value.
+  useEffect(() => {
+    if (!isEqual(initializedConfig, config) && onConfigChange) {
+      onConfigChange(initializedConfig);
+    }
+  }, [initializedConfig, config, onConfigChange]);
+
+  return (
+    <StylesProvider generateClassName={generateClassName}>
+      <ThemeProvider theme={muiTheme[theme]}>
+        <DatasetLoaderProvider>
+          <VitessceGrid
+            config={initializedConfig}
+            getComponent={getComponent}
+            rowHeight={rowHeight}
+            height={height}
+            theme={theme}
+            onWarn={onWarn}
+          />
+          <ViewConfigPublisher
+            onConfigChange={onConfigChange}
+            onLoaderChange={onLoaderChange}
+          />
+        </DatasetLoaderProvider>
+      </ThemeProvider>
+    </StylesProvider>
+  );
+}
+
+/**
+ * The Vitessce component.
+ * @param {object} props
+ * @param {object} props.config A Vitessce view config.
+ * If the config is valid, the VitessceGrid will be rendered as a child.
+ * If the config is invalid, a Warning will be rendered instead.
+ * @param {number} props.rowHeight Row height for grid layout. Optional.
+ * @param {number} props.height Total height for grid layout. Optional.
+ * @param {string} props.theme The theme, used for styling as
+ * light or dark. Optional. By default, "dark"
+ * @param {function} props.onWarn A callback for warning messages. Optional.
+ * @param {function} props.onConfigChange A callback for view config
+ * updates. Optional.
+ * @param {function} props.onLoaderChange A callback for loader
+ * updates. Optional.
+ */
+export default function Vitessce(props) {
+  const {
+    config,
+    theme,
+  } = props;
+  if (!config) {
+    // If the config value is undefined, show a warning message.
+    return (
+      <Warning
+        title="No such dataset"
+        unformatted="The dataset configuration could not be found."
+        theme={theme}
+      />
+    );
+  }
+  // If the view config is missing a version, show a warning message.
+  if (!config.version) {
+    return (
+      <Warning
+        title="Missing version"
+        unformatted="The dataset configuration is missing a version, preventing validation."
+        theme={theme}
+      />
+    );
+  }
+  // Check if this is a "legacy" view config.
+  let upgradedConfig = config;
+  if (config.version === '0.1.0') {
+    // Validate under the legacy schema first.
+    const validateLegacy = new Ajv().compile(legacyConfigSchema);
+    const validLegacy = validateLegacy(config);
+
+    if (!validLegacy) {
+      const failureReason = JSON.stringify(validateLegacy.errors, null, 2);
+      return (
+        <Warning
+          title="Config validation failed"
+          preformatted={failureReason}
+          theme={theme}
+        />
+      );
+    }
+    // Upgrade from v0.1.0 to v1.0.0 before v1.0.0 schema validation.
+    upgradedConfig = upgrade(config);
+  }
```

Comment:
I wonder if this is the issue with https://github.com/hubmapconsortium/vitessce/pull/728#issuecomment-683830818 since `upgrade` creates a new config object

Reply:
One option could be to cache view states, but I don't know what that would look like -  hashing a string representation of the object?

Reply:
Oh I see, so the state is cleared every time the `<Vitessce/>` component re-renders, which may occur if the `height` prop changes (when going to full screen mode).

I think a solution would be to store the original view config prop in state or a ref and then diff against it during each render to determine whether it has updated. `useMemo` might work for this actually

Reply:
Good point - we don't need to be aware of the history of past configs seen.  `useMemo` seems like a good solution (or something similar like you mentioned)